aten/src/ATen/core/ivalue.cpp

```diff
@@ -297,6 +298,46 @@ IValue IValue::equals(const IValue& rhs) const {
   TORCH_INTERNAL_ASSERT(false, "we should never reach here")
 }
 
+size_t IValue::hash(const IValue& v) {
+  switch (v.tag) {
+    case Tag::None:
+      return 0;
+    case Tag::Bool:
+      return c10::get_hash(v.payload.as_bool);
+    case Tag::Double:
+      return c10::get_hash(v.payload.as_double);
+    case Tag::Tensor:
+      // Tensor __hash__ is equivalent to `id()`, so take the pointer value of
+      // the tensor to emulate it
+      return c10::get_hash(v.payload.as_int);
```

Comment:
I don't think this violates the guarantees for `hash` but I just want to point out that we're interpreting an unsigned integer (`as_intrusive_ptr`) as a signed integer. This usually leads to problems but I think it should be okay in this case?

Reply:
Good catch, I think that's fine here given that number of distinct values is still the same, hash will be just as good.